### PR TITLE
Implemented a base type routing convention and updated examples to de…

### PR DESCRIPTION
**Pull request by JonyTek**
Hi Dasith,

I have just pushed the work that I did last night, there are now merge conflicts that I can resolve if you think some of this is useful.

I did do something very similar with the Swashbuckle work with setting the HttpVerb.

One cool thing I've done is removing the need to have the HttpVerb attributes with the implementation of the marker interfaces on the verb scoped base handlers.

Please let me know your thoughts.

### CODE REVIEW

**Comment by dasiths** [comment]
Hey Jonathan,
I love the changes. The only suggestion I would make is to check whether there is existing `HttpMethodMetadata` before adding the new one.

I think your controller metadata option is much cleaner than my choice of using action attributes. We can get rid of the need for the extra package I introduced. Should have waited for yours. Well done.

One suggestion though. Since we are now using configuration settings on setting up can we use DI to inject it in? So both `MvcOptions` and  `SimpleEndpointsConfiguration` be done this way.

Something like https://github.com/aspnet/Mvc/issues/6214#issuecomment-298133612

Please feel free to remove the Swashbuckle package I added. You might need to update the `build.yaml` and `ReadMe.md` but they are one liners. I'm happy to help out.

**Comment by JonyTek** [comment]
Ok great! I'm glad you like the ideas. 

> The only suggestion I would make is to check whether there is existing HttpMethodMetadata before adding the new one.

Yes we can do this.

> I think your controller metadata option is much cleaner than my choice of using action attributes.

I assume you're referring to the use of the marker interfaces here? I think it's a neat solution.

> One suggestion though. Since we are now using configuration settings on setting up can we use DI to inject it in? So both MvcOptions and SimpleEndpointsConfiguration be done this way.

Just to clarify on this you're thinking about moving the `.AddSimpleEndpoints()` registration up onto I`IServiceCollection` and off the `MvcOptions` object? I can look at this but have a feeling it will need us to reference another nuget package. 

I will go ahead and make the changes and sort out the merge conflict, Ill also get some tests in there today.

**Comment by dasiths** [comment]
> I assume you're referring to the use of the marker interfaces here? I think it's a neat solution.

I was referring to the `EndpointMetadata` in the controller you're using vs the `ActionConstraints` I had been using. I think doing everything in the controller level is probably better.

>Just to clarify on this you're thinking about moving the .AddSimpleEndpoints() registration up onto IIServiceCollection and off the MvcOptions object? I can look at this but have a feeling it will need us to reference another nuget package.

`IMvcBuilder` comes from `Microsoft.Extensions.DependencyInjection` which should already be part of the dependencies for `Microsoft.AspNetCore.Mvc.Core`

```
    public class EndpointRoutingConvention: IApplicationModelConvention
    {
        private const string EndpointString = "endpoint";
        private readonly EndpointRoutingConfig _endpointRoutingConfig;

        public EndpointRoutingConvention(IOptions<EndpointRoutingConfig> endpointRoutingConfig)
        {
            _endpointRoutingConfig = endpointRoutingConfig.Value;
        }

        public virtual void Apply(ApplicationModel application)
        {
            // use _endpointRoutingConfig here
        }
    }

    public class EndpointRoutingConfig
    {
           // your properties here
    }

    public class ConfigureMvcOptions: IConfigureOptions<MvcOptions>
    {
        private readonly EndpointRoutingConvention _endpointRoutingConvention;

        public ConfigureMvcOptions(EndpointRoutingConvention endpointRoutingConvention)
        {
            _endpointRoutingConvention = endpointRoutingConvention;
        }
        public void Configure(MvcOptions options)
        {
            options.Conventions.Add(_endpointRoutingConvention);
        }
    }

// Then in the extension

        public static IMvcBuilder AddSimpleEndpointsConventions(this IMvcBuilder mvcBuilder, Action<EndpointRoutingConfig> setupAction = null)
        {
            mvcBuilder.Services.AddSingleton<EndpointRoutingConvention>();
            mvcBuilder.Services.AddSingleton<IConfigureOptions<MvcOptions>, ConfigureMvcOptions>();

            if (setupAction != null)
            {
                mvcBuilder.Services.Configure<EndpointRoutingConfig>(setupAction);
            }

            return mvcBuilder;
        }
```

finally in startup.cs
```
services.AddControllers().AddSimpleEndpointsConventions();
```

File: src/SimpleEndPoints/Conventions/EndpointRoutingConvention.cs [hunk 1]
```diff
@@ -1,22 +1,77 @@
...
+                MapRouteConvention(controller);
+            }
+        }
+
+        private  void MapRouteConvention(ControllerModel controller)
+        {
+            var routeTemplate = controller.Selectors[0].AttributeRouteModel.Template;
+
+            //If we haven't set a route with [Route("")] on our endpoint
+            if (routeTemplate.Equals(EndpointPlaceholder, StringComparison.OrdinalIgnoreCase))
```
Comment (dasiths) [inline src/SimpleEndPoints/Conventions/EndpointRoutingConvention.cs]: Should this not be contains() to cater for `[Route("custom/[endpoint]/another")]`

File: src/SimpleEndPoints/Conventions/RouteMutator.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    internal class RouteMutator : IConventionMutator
+    {
+        private const string EndpointPlaceholder = "[endpoint]";
+
+        public void Mutate(ControllerModel controller, SimpleEndpointsConfiguration configuration)
+        {
+            var routeBuilder = new StringBuilder();
+            var routeTemplate = controller.Selectors[0].AttributeRouteModel.Template;
+
+            if (routeTemplate.Equals(EndpointPlaceholder, StringComparison.OrdinalIgnoreCase))
```
Comment (dasiths) [inline src/SimpleEndPoints/Conventions/RouteMutator.cs]: Should this not be .indexof(blah, ignorecase) >= 0?

**Comment by JonyTek** [comment]
Hi buddy, so I did make some updates and resolved the merge issues. I'm not sure I have gone down the path you were recommending in your comments but wanted to commit anyway.

Please comment in the code for things that you don't agree with just to help me better understand your points.

In terms of the dependency registrations I did go a slightly different path and it was written before I saw your last comment so I do apologise. 

I have added some tests, both unit and integration. I can also update the README if you would be happy to merge.

Thanks.

**Comment by dasiths** [comment]
@JonyTek 

Shall I pull down your branch and do the changes if it's easier for you? You're 90% there.

File: src/SimpleEndPoints/Conventions/EndpointRoutingConvention.cs (hunk 1 above)
Comment (JonyTek) [reply src/SimpleEndPoints/Conventions/EndpointRoutingConvention.cs]: It did occur to me that this was your intention originally and i had removed it without knowing, a cool feature that I will add back in.

**Comment by JonyTek** [comment]
That would be great mate.

**Comment by dasiths** [comment]
@JonyTek what do you think of the changes I made?

**Comment by dasiths** [comment]
I'll merge this in but feel free to create a PR for improvements.

**Comment by dasiths** [comment]
Thanks again for the wonderful suggestions and ideas.

**Comment by JonyTek** [comment]
Awesome mate! Really happy to help. Do you have any thoughts on additional features? Im in London and figure that I am going to have a fair amount of time on my hands over the coming weeks/months.

**Comment by dasiths** [comment]
I'm open to suggestions. How about extending your mutator idea by allowing for developers to inject their own?

**Pull request merged**